### ISSUE TITLE
The state interferes with printing expression(s)

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#4703]
Stop the state from interfering with the printing (press)!

>> However, in general this means that debug printing (via prettyTCM) may interfere with ordinary printing?

> Yes, that's an unfortunate (?) side effect of the new way I implemented the shadowing. Since it is only a problem when using verbosity levels, I didn't think it was worth fixing. However, we could conceivably wrap every call to `reportSDoc` in a `localTCState` if it becomes a problem. (or something more fine-grained that just restores the shadowing information)

_Originally posted by @jespercockx in https://github.com/agda/agda/pull/4699#issuecomment-636828338_

**Comment by jespercockx** [comment]
In general, I do not believe having a certain degree of statefulness in the pretty-printer can be avoided in the current design of Agda. We quite frequently print terms containing different free variables that happen to have the same name, so we need to rename one or the other to disambiguate. However, the current way in which we print error and interactive messages does not allow us to make this decision globally. In fact, in the case of interaction we might already have committed to a certain name so we *have* to rely on the state of the set of variable names we already used.

In the case of debug printing specifically, it is indeed surprising that enabling some debug prints can change how variables are named. However, since it does not affect the experience of the typical user I did not think it was worth fixing (as I said before in my comment). If it bothers you, I can fix it.

**Comment by nad** [comment]
Perhaps we could take the approach of Epigram 2, where you can use de Bruijn indices to refer to shadowed names. I don't remember the exact syntax, but I think it was something like the following: `(A : Type) (A x y : A) → A^1`. Here `A^1` means "the second thing called `A`".

**Comment by rwe** [comment]
Hi, so this might be relevant to something I'm looking at. I'm not an expert so I'd appreciate your feedback on this…

As part of looking back into ways to decouple some of the Haskell modules, I came to the realization that `PrettyTCM` is the stickiest part of all that coupling: obviously we can't get rid of debug messages, but `MonadPretty`/`PrettyTCM` basically requires the kitchen sink, even though most things don't actually need *everything* to just define or use pretty printing. For some things, it's the biggest source of dependencies on `Agda.TypeChecking.Monad.Base`, which in turn is the biggest dependency focal point.

That's an issue for type restriction (i.e. you _cannot_ define or use `prettyTCM` without also requiring a very powerful monad). It's also a major dependency issue, since the _definition_ of `MonadPretty`/`MonadTCM` refers to all those things.

For reference, here's the definition of `PrettyTCM`:
```hs
class PrettyTCM a where
  prettyTCM :: MonadPretty m => a -> m Doc
```

Where `MonadPretty m` ultimately expands out to these constraints:

```hs
type MonadPretty m =
  ( Fail.MonadFail m
  , MonadAddContext m
  , MonadFresh NameId m
  , MonadInteractionPoints m
  , HasBuiltins m
  , HasConstInfo m
  , HasOptions m
  , MonadAbsToCon m
  , MonadDebug m
  , MonadReduce m
  , MonadReify m
  , MonadTCEnv m
  , ReadTCState m
  , Applicative m, Functor m, Monad m
  , Null (m Doc), Semigroup (m Doc), IsString (m Doc)
  )
```

I'm toying with parameterizing that typeclass on the result monad, so those instances can be defined a little more abstractly. On the surface the change is very minor.

Something like:

```hs
class PrettyM m a where
  prettyM :: a -> m Doc
```

As an example instance:
```hs
instance ( Applicative m, IsString (m Doc)
         , PrettyM m Type, PrettyM m a
         ) => PrettyM m (Judgement a) where
  prettyM (HasType a cmp t) = prettyM a <+> ":" <+> prettyM t
  prettyM (IsSort  a t)     = "Sort" <+> prettyM a <+> ":" <+> prettyM t
```
(`Applicative`, `IsString` etc could be harmlessly bundled, but note that not all instances even need those).

Which allows a lot more focused constraints. Ultimately the monad will usually be `TCM`, but this way fewer modules are forced to have direct knowledge of that, depending on what concrete types they're working with.

My hope is that this will make it easier to decouple things from literal references to the typechecking monad, so that modules can be built a bit more independently, and in turn make it easier to restrict some of the state access that might contribute to problems like this issue.

This comes at the cost of more instance-solving by GHC, but I hope it would be more-than-offset by gains from decoupling the dependency web. At first it might seem like it would come at the cost of specialization: but since everything was already parameterized on `MonadPretty` (and, at the method level, not the instance level), I don't think that it will be worse there.

@asr @andreasabel @jespercockx @gallais does that seem like a reasonable thing to do? Is there another approach in mind or already in the works?

**Comment by jespercockx** [comment]
In previous times, `prettyTCM` used to have type `a -> TCM Doc`, before I introduced the `MonadPretty` constraint to make the dependencies more explicit. My main goal was to statically rule out *modifications* to the state during pretty-printing by only using `ReadTCState` rather than `MonadTCState`, hence I just used the same constraint `MonadPretty` everywhere. However, all efforts into further reducing the constraints that specific pretty-printing functions rely on would be very welcome. The approach you describe also makes sense to me.

tldr: PR's welcome ;)

**Comment by rwe** [comment]
An issue I've run into with the approach I described above: in many cases it requires `UndecidableInstances`, which is fairly infective. I'm looking into other ways of doing the same thing without proliferating special-cased classes or making anything too invasive. (Should I turn that into a separate issue to avoid cluttering this bug report?)

**Comment by rwe** [comment]
By the way: I did actually get that working mostly non-invasively without additional `UndecidableInstances`, though in several cases it required hacks to coax the instance resolution (parameterizing some of the data types).

The undecidable instance issues aren't entirely caused by the proposed class structure. It's often caused by the existing constraints which are pulled up from their original definition sites (e.g. `TypeChecking/Monad/Base.hs` and `TypeChecking/Monad/Pretty.hs`)—which already do have `UndecidableInstances` turned on—to the use sites, which may not.

There are a few sources of those:
  1. `MonadFresh NameId m`, incurred by `MonadReify m`. The multiparam typeclass without unwrapping either parameter in the constraint makes GHC sad, especially due to its indirect instantiation via `HasFresh NameId`. This could probably solved by creating a separate `MonadFreshNameId m` class.
  2. An implicit chain of instance requirements for pretty-reifying internal syntax: `(Reify i a, ToConcrete a c, Pretty c) => i → Doc`. Fundeps tell GHC that `i` determines `a` and `a` determines `c`, but stating that in an instance declaration requires `QuantifiedConstraints` (GHC 8.6.1). This is not a huge problem, though, because there aren't actually *that* many relevant instances, so can be fixed by explicit instantiation.
  3. `IsString (m Doc)`/`Null (m Doc)`/`Semigroup (m Doc)` caused by the implicit lifting. I think this is evidence that the pretty printing ought to happen in its own monad or be a transformer. (The `Semigroup` operation really refers to combining pretty-serialization data, which may happen to use the TCM; not vice-versa. I'd guess this should really be something like `a → DocT m` instead of `a → m Doc`, and that _that_ is the main point relevant to the original issue description).

Finally, constraints like `instance (PrettyM m Term) => PrettyM m SplitPatVar` typically requires `UndecidableInstances` regardless. The error both that and the `IsString (m Doc)` stuff is like:
> ```
> The constraint ‘PrettyM m Term’
> is no smaller than the instance head ‘PrettyM m SplitPatVar’
> ```

This can be worked around by adding type parameters to the types , for example:
```diff
- data SplitPatVar = …
+ type SplitPatVar = SplitPatVar' Term
+ data SplitPatVar' ter = …
 
  instance (PrettyM m Term) => PrettyM m SplitPatVar where …
```
Which can be useful in some cases anyway, but in this case has a questionable justification. (I don't _really_ understand why that fixes it without otherwise touching the instance declaration).

It can also be worked around by adding sufficient constraints so that the dependent instances are resolvable at the point of definition (basically what `PrettyTCM` currently does): though this kind of mucks with the process of decoupling them. It also doesn't solve the `IsString (m Doc)` constraints when they're needed: those nearly _always_ need `UndecidableInstances`.

Or by parameterizing the instances by a data-family that gets unwrapped (e.g. `instance (PrettyM n Foo m) => PrettyM (Succ n) Bar m`), which works, though is unusual, brittle, and requires some way to communicate that type parameter at call sites.

Or by turning on `UndecidableInstances` in more places, which is frequently suggested online but also seems nuclear. Not sure what the right fix would be.
